### DISCUSSION TITLE
How to mount a component using vanilla Javascript?

### DISCUSSION THREAD

**Discussion by AlexErrant** [solidjs/solid#1755] [Q&A] [answered]
...while keeping reactivity, of course.

(See [this discussion](https://github.com/solidjs/solid/discussions/1751) for why I want to use vanilla Javascript - basically trying to support plugins written in plain Javascript.)

I've tried a few ways, none of which have worked.

## Setup:

```ts
const Comp: VoidComponent<{
  i: number // some reactive prop
  child: VoidComponent<{ i: number }> // some Solid component passed to Javascript that I'm trying to mount
}> = (props) => {
  const container = document.createElement("div")
  // insert attempted code from below here
  return container // this Node is mounted in a solid-app here: https://github.com/AlexErrant/solid-plugin/blob/56729963cc6b3dacc760e123a43e30d5ad46d55e/my-solid-project/src/App.tsx#LL35C16-L35C36
}
export default Comp
```

I don't mind importing `VoidComponent` from Solid because it's just a type definition (really just a function returning an Element/Node). Heck, I don't mind importing anything really from Solid, including:

## Solid's `render`:
```ts
  const renderDiv = document.createElement("div")
  render(() => props.child({ i: props.i }), renderDiv)
  container.appendChild(renderDiv)
```

However, it fails to update when `i` changes.

## Solid's `Dynamic`:
```ts
  // @ts-expect-error Dynamic is callable, contrary to the typescript definition. Calling it is probably breaking reactivity, but I don't know how else to attach it to the DOM.
  const dynamicNode = Dynamic({ component: props.child, i: props.i })() as Node
  container.appendChild(dynamicNode)
```

It also fails to update when `i` changes.

## Solid's `Portal`:
```ts
  const portalDiv = document.createElement("div")
  Portal({
    mount: portalDiv,
    children: props.child({ i: props.i }),
  })
  container.appendChild(portalDiv)
```

It too fails to update when `i` changes.

## Invoking the component
...and replacing the entire DOM sub-tree with createEffect. However, this loses all fine-grained updates:

```ts
  let componentDiv = document.createElement("div")
  createEffect(() => {
    componentDiv.replaceChildren(props.child({ i: props.i }) as Node)
  })
  container.appendChild(componentDiv)
```

I'm out of ideas on what else to try. Here's [a minimal repo](https://github.com/AlexErrant/solid-plugin), here's [the Typescript plugin](https://github.com/AlexErrant/solid-plugin/blob/main/ts-plugin/src/counter.ts) attempting to use the above, and finally here's [the Stackblitz](https://stackblitz.com/~/github.com/AlexErrant/solid-plugin) demoing the lack of reactivity propagation. Note that if you want to play with the Stackblitz, you'll need to modify the [`tsScript.ts`](https://github.com/AlexErrant/solid-plugin/blob/main/my-solid-project/src/tsScript.ts), as that's where the compiled Javascript is loaded from. You'll also need to `pnpm build && pnpm serve` due to [an issue with Vite](https://github.com/vitejs/vite/issues/6582).

Edit: don't use the Stackblitz, I made a playground repro: https://playground.solidjs.com/anonymous/16aa5f90-9494-4f0e-ab9f-005fdf20429c

**Comment by atk** [comment, marked as answer]
You need to use `createComponent from "solid-js"` to retain reactivity, e.g.:

```ts
createComponent(Dynamic, { ... });
```

**Reply by AlexErrant** [reply]
Hm, I tried with a plain component:

```ts
const createPlainComponent = createComponent(props.child, { i: props.i })
container.appendChild(createPlainComponent as Node)
```

and with a `Dynamic` component:

```ts
  const createDynamicComp = createComponent(Dynamic, {
    component: props.child,
    i: props.i,
  })
  // @ts-expect-error Dynamic is callable, contrary to the typescript definition. Calling it is probably breaking reactivity, but I don't know how else to attach it to the DOM.
  container.appendChild(createDynamicComp() as Node)
```

but neither retained reactivity.

I also tried combining `createEffect` with `Portal`, but it too loses fine-grained reactivity:

```ts
  const portalEffectDiv = document.createElement("div")
  createEffect(() => {
    Portal({
      mount: portalEffectDiv,
      children: props.child({ i: props.i }),
    })
  })
  container.appendChild(portalEffectDiv)
```

Here's [the commit](https://github.com/AlexErrant/solid-plugin/commit/56729963cc6b3dacc760e123a43e30d5ad46d55e#diff-807978bc5ae50fa2427b9e6bdb7bb8c388ad0205d1ef552a7fb527454fe2d746). I'm open to more ideas - please don't feel compelled to try [the (now fixed) Stackblitz](https://stackblitz.com/~/github.com/AlexErrant/solid-plugin) yourself!

**Reply by AlexErrant** [reply]
Selecting this as the answer; details at the bottom of the page!

**Comment by atk** [comment]
Why are you avoiding `render`?

**Reply by AlexErrant** [reply]
Hm, but I did try `render`? It's right below `Setup` in my OP:

```ts
  const renderDiv = document.createElement("div")
  render(() => props.child({ i: props.i }), renderDiv)
  container.appendChild(renderDiv)
```

Here's [the actual code](https://github.com/AlexErrant/solid-plugin/blob/main/ts-plugin/src/counter.ts#L44-L48), and here's [the Stackblitz](https://stackblitz.com/~/github.com/AlexErrant/solid-plugin) showing that `render` doesn't update on prop change. It's the first option under the big "_Failing attempts to attach a Solid Component_"

Note that the `createEffect` attempts do update on prop change, so this is still in a reactive context; however, I don't know how to make the dom updates from `createEffect` fine-grained.

**Reply by AlexErrant** [reply]
I created a minimal repo on the playground demoing the various attempts I've tried: https://playground.solidjs.com/anonymous/16aa5f90-9494-4f0e-ab9f-005fdf20429c The lack of reactivity is apparent, as is the lack of fine-grained updates with `createEffect`

I removed all the `importmap`/`import()`/gnarly plugin code, so now I just have a function (`Plugin`) that returns a simple `VoidComponent`, so this even more minimal than my Stackblitz.

ESLint is telling me

> The reactive variable 'props.i' should be used within JSX, a tracked scope (like createEffect), or inside an event handler function, or else changes will be ignored. eslint(solid/reactivity)

However, since _the point_ of a vanilla javascript plugin is that it's vanilla, I don't have JSX. `createEffect` isn't fine-grained - perhaps there's another way to put it in a tracked scope? And finally I've no idea how to put it in an event handler function 😅

**Reply by atk** [reply]
I didn't see any attempt that uses `render` together with `createComponent`, very much like our testing-library does: https://github.com/solidjs/solid-testing-library/blob/main/src/index.ts#L87

**Reply by AlexErrant** [reply]
Unfortunately, I can't get `render` + `createComponent` to work. My naive impl:

```ts
  const renderDiv = document.createElement("div");
  render(() => {
    return createComponent(props.child, { i: props.i });
  }, renderDiv);
  container.appendChild(renderDiv);
```

This doesn't update at all and gives a lint warning similar to the one in the above comment.

Things are better if I use the wrapper from the solid-testing-library code:

```ts
  const wrapper: Component<{ children: JSX.Element }> = (props) => (
    <div>Wrapper {props.children}</div>
  );
  const wrappedUi = () =>
    createComponent(wrapper, {
      get children() {
        return createComponent(props.child, { i: props.i });
        // return props.child({ i: props.i }); // I also tried this, but same problem
      },
    });
  const wrapperRenderDiv = document.createElement("div");
  render(wrappedUi, wrapperRenderDiv);
  container.appendChild(wrapperRenderDiv);
```

This has no lint warning and _does_ update, but they aren't _fine-grained_ updates; it updates the entire component's DOM.

[New playground of just this attempt.](https://playground.solidjs.com/anonymous/8e9a9bcc-220c-4f64-a874-bf8bb8afa606)

By the way, I really appreciate your help and ideas! This is a lot better than me going through the API documentation going "hmm, maybe this primitive?"

**Reply by atk** [reply]
I'll have a look later about those fine-grained updates.

**Reply by AlexErrant** [reply]
Additionally, I think the `wrappedUi` works because it uses JSX, which is something I want to avoid in a vanilla Javascript solution. When I remove the JSX from the wrapper...

```ts
const wrapper: Component<{ children: JSX.Element }> = (props) => props.children;
```

I get an `eslint(solid/reactivity)` warning on `props.children`, and reactivity fails to propagate.

I don't mind calling Solid functions, but requiring plugin authors to compile Solid JSX with a build step when they're writing plain ol' Javascript DOM manipulation isn't ideal.

**Reply by atk** [reply]
Just put the JSX in the playground and use the output.

**Comment by snnsnn** [comment]
Your intent is still not clear to me but here is an example that uses plain javascript both on the root and the leaf.

```tsx
import {
  Component,
  JSXElement,
  createComponent,
  createEffect,
  createMemo,
  createRoot,
  createSignal,
  onCleanup,
} from 'solid-js';
import { render } from 'solid-js/web';

const Comp: Component<{ count: number }> = (props) => {
  let div = document.createElement('div');
  createEffect(() => {
    div.innerHTML = `Count: ` + props.count;
  });
  return div;
};

const App = () => {
  let dispose: () => void;
  const [count, setCount] = createSignal(0);
  setInterval(() => setCount((p) => p + 1), 1000);

  onCleanup(() => {
    dispose && dispose();
  });

  const el = createRoot((disposer) => {
    dispose = disposer;
    return createMemo(() =>
      createComponent(Comp, {
        get count() {
          return count();
        },
      }),
    );
  }) as unknown as JSXElement;

  return <div>{el}</div>;
};

render(App, document.body);
```

**Reply by AlexErrant** [reply]
Sorry, but this isn't quite it either. The goal is to mount a Solid _component_ using only Javascript. Yes, I can get a reactive signal down to the leaf, but that's not a _Component_, [by Solid's definition](https://github.com/solidjs/solid/blob/5e2999e4faffa5d8b9f676bd8d3b230eee5a1ad4/packages/solid/src/render/component.ts#L23):

```ts
export type Component<P = {}> = (props: P) => JSX.Element;
```

Basically, a Component is a function that returns a `Node | ArrayElement | (string & {}) | number | boolean | null | undefined`; in my examples I've been returning an `HTMLDivElement` (which satisfies `Node`) which I've called [`container`](https://github.com/AlexErrant/solid-plugin/blob/56729963cc6b3dacc760e123a43e30d5ad46d55e/ts-plugin/src/counter.ts#L18).

I tweaked your code to pass a Solid component through to the leaf. Please let me know how you'd attach it to the DOM:

```ts
import {
  type VoidComponent,
  type Component,
  type JSX,
  createComponent,
  createEffect,
  createMemo,
  createRoot,
  createSignal,
  onCleanup,
} from "solid-js";
import { render } from "solid-js/web";

const ChildComp: VoidComponent<{ i: number }> = (props) => {
  return <div>Child Comp, *2 prop is {props.i * 2}</div>
}

const Comp: Component<{ count: number, child: VoidComponent<{ i: number }> }> = (props) => {
  let div = document.createElement("div");
  createEffect(() => {
    div.innerHTML = `Count: ` + props.count;
    // how do I attach props.child to the DOM here in a reactive way?
  });
  return div;
};

const App = () => {
  let dispose: () => void;
  const [count, setCount] = createSignal(0);
  setInterval(() => setCount((p) => p + 1), 1000);

  onCleanup(() => {
    dispose && dispose();
  });

  const el = createRoot((disposer) => {
    dispose = disposer;
    let r = createMemo(() =>
      createComponent(Comp, {
        get count() {
          return count();
        },
        get child(){
          return ChildComp;
        }
      })
    );
    return r;
  }) as unknown as JSX.Element;

  return <div>{el}</div>;
};

render(App, document.body);
```

To elaborate a bit on _why_ I'm trying to do this, I'm writing an app in Solid that supports plugins. Plugins (ideally) shouldn't be in only Solid; they should also support plain Javascript. So assume I have a plugin written in Javascript that takes as an argument a Solid component. How would the Javascript-plugin attach that component to the DOM without the benefit of JSX?

In [my playground link](https://playground.solidjs.com/anonymous/16aa5f90-9494-4f0e-ab9f-005fdf20429c), `Counter` is the primary app, `Plugin` is the plain vanilla Javascript plugin (ideally without JSX), and `ChildComp` is the Solid component that `Counter` is passing to the `Plugin` for it to mount somewhere in its own DOM. `Plugin` returns a DOM element (with `ChildComp` in it somewhere) that `Counter` renders with `<Dynamic>`. Here's an HTML wireframe:

![image](https://github.com/solidjs/solid/assets/109672176/2d76ad7e-13ea-4d9d-b9dd-f6da375f75e6)

`Counter` defines `ChildComp`, which it passes to `Plugin` to render wherever it likes. I hope this makes sense!

**Reply by atk** [reply]
Did you make sure that the same instance of solid is used on all levels?

**Comment by snnsnn** [comment]
You know Comp is an actual component and it is reactive. I use imperative API because you were using it almost in all your examples. 

```tsx
import {
  Component,
  JSXElement,
  createComponent,
  createEffect,
  createMemo,
  createRoot,
  createSignal,
  onCleanup,
} from 'solid-js';
import { render } from 'solid-js/web';

const Comp: Component<{ count: number }> = (props) => {
  return (
    <div>
      <div>Count: {props.count}</div>
    </div>
  );
};

const App = () => {
  let dispose: () => void;
  const [count, setCount] = createSignal(0);
  setInterval(() => setCount((p) => p + 1), 1000);

  onCleanup(() => {
    dispose && dispose();
  });

  const el = createRoot((disposer) => {
    dispose = disposer;
    return createComponent(Comp, {
      get count() {
        return count();
      },
    });
  });

  return <div>{el}</div>;
};

render(App, document.body);
```

>  // how do I attach props.child to the DOM here in a reactive way?

I don't get this, please refactor the code in way to achieve your purpose, we can fix the errors and work towards your goal.

Do you want to render props.children reactively?

```tsx
import {
  Component,
  createComponent,
  createRoot,
  createSignal,
  onCleanup
} from 'solid-js';
import { render } from 'solid-js/web';

const Comp: Component<{ children: number }> = (props) => {
  return () => props.children;
};

const App = () => {
  let dispose: () => void;
  const [count, setCount] = createSignal(0);
  setInterval(() => setCount((p) => p + 1), 1000);

  onCleanup(() => {
    dispose && dispose();
  });

  const el = createRoot((disposer) => {
    dispose = disposer;
    return createComponent(Comp, {
      get children() {
        return count();
      },
    });
  });

  return <div>{el}</div>;
};


render(App, document.body);
```

**Reply by snnsnn** [reply]
You don't need a separate root but it is just to show you how you can create one.

```tsx
import {
  Component,
  createComponent,
  createSignal
} from 'solid-js';
import { render } from 'solid-js/web';

const Comp: Component<{ children: number }> = (props) => {
  return () => props.children;
};

const App = () => {
  const [count, setCount] = createSignal(0);
  setInterval(() => setCount((p) => p + 1), 1000);

  const el = createComponent(Comp, {
      get children() {
        return count();
      },
  });

  return <div>{el}</div>;
};

render(App, document.body);
```

**Comment by AlexErrant** [comment]
Okay, I think I figured it out, please no one try to grok my shitty technical writing. I'll post my solution in a bit, gotta step away to emotionally recover from my stupidity.

**Reply by AlexErrant** [reply]
Okay @atk's comment

> Just put the JSX in the playground and use the output.

finally sank in. _How about I write the Plugin in Solid, then look at the compiled output and go from there?_ So, a Solid plugin like

```ts
const Plugin: VoidComponent<{
  i: number;
  child: VoidComponent<{ i: number }>;
}> = (props) => {
  return <Dynamic component={props.child} i={props.i}/>;
};
```

looks like this in Javascript:


```ts
const Plugin = props => {
  return _$createComponent(Dynamic, {
    get component() {
      return props.child;
    },
    get i() {
      return props.i;
    }
  });
};
```

Which is _exactly_ what @atk initially suggested. However, I implemented his suggestion like this

```ts
const createPlainComponent = createComponent(props.child, { i: props.i })
container.appendChild(createPlainComponent as Node)
```

_...which is missing getters._

Lo and behold, [when I update my repo](https://github.com/AlexErrant/solid-plugin/commit/cb405eeaa5f9c8f7a084f0269de8d05059cb1035)'s attempts to use getters, literally _every_ attempt works, and _all_ are fine-grained. 🤦 

Thanks so much @atk for your help and patience!
